web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -0,0 +1,179 @@
+# PDEP-6: Ban upcasting in setitem-like operations
+
+- Created: 23 December 2022
+- Status: Under discussion
+- Discussion: [#50402](https://github.com/pandas-dev/pandas/pull/50402)
+- Author: [Marco Gorelli](https://github.com/MarcoGorelli) ([original issue](https://github.com/pandas-dev/pandas/issues/39584) by [Joris Van den Bossche](https://github.com/jorisvandenbossche))
+- Revision: 1
+
+## Abstract
+
+The suggestion is that setitem-like operations would
+not change a ``Series``' dtype.
+
+Current behaviour:
+```python
+In [1]: ser = pd.Series([1, 2, 3], dtype='int64')
+
+In [2]: ser[2] = 'potage'
+
+In [3]: ser  # dtype changed to 'object'!
+Out[3]:
+0         1
+1         2
+2    potage
+dtype: object
+```
+
+Suggested behaviour:
+
+```python
+In [1]: ser = pd.Series([1, 2, 3])
+
+In [2]: ser[2] = 'potage'  # raises!
+---------------------------------------------------------------------------
+TypeError: Invalid value 'potage' for dtype int64
```

Comment:
I think we want a TypeError as there is no value of this type that would be valid. But I would be +1 on the message saying something like "Invalid type {type(value)} for dtype..."

Reply:
> as there is no value of this type that would be valid

That's actually a good question: is that the case? Would `ser[2] = "10"` also fail? 
In numpy that doesn't (and it casts the string to an int), but pandas doesn't do that currently and preserves the string values and upcasts to object dtype. I assume we want to raise an error instead of silently trying to cast (that also matches the current PDEP text, since it is a case that currently upcasts to object dtype). 

But if we take a different data type, datetimelike instead of numeric, we actually _do_ try to cast the string value. That's actually used as first example in the motivation section here:

```python
In[9]: ser = pd.Series(pd.date_range("2000", periods=3))
In[10]: ser[2] = "2000-01-04"  # works, is converted to datetime64
In[11]: ser[2] = "2000-01-04x"  # typo - but pandas does not error, it upcasts to object
```

In this case I assume it would be a ValueError instead of TypeError. But exact error classes aside (that's a detail that doesn't need to be fixed in this PDEP), it _is_ maybe a relevant question whether we want to treat such string casting more consistently. 
Although I know that in general we always have been liberal in allowing to use strings instead of Timestamp literals in various places, eg also in indexing (where we certainly do _not_ want to automatically interpret a string as an integer), so there is precedent in doing this differently for datetime-like dtypes, and thus this is probably better for a separate issue.

Reply:
numpy uses ValueError for this, so for now I've gone with that